addons/stock_picking_batch/models/stock_move_line.py

```diff
@@ -88,3 +89,178 @@ def _add_to_wave(self, wave=False):
         if picking_to_wave_vals_list:
             self.env['stock.picking'].create(picking_to_wave_vals_list)
         wave.action_confirm()
+
+    def _find_auto_wave(self):
+        """Try to find compatible waves to attach the move lines to, if wave grouping is enabled."""
+        lines_nearest_parent_locations = defaultdict(lambda: self.env['stock.location'])
+        for line in self:
+            if line.picking_type_id.wave_group_by_location:
+                wave_locations = line.picking_type_id.wave_location_ids.sorted(key=lambda l: len(l.complete_name), reverse=True)
+                for location in wave_locations:
+                    if line.location_id._child_of(location):
+                        lines_nearest_parent_locations[line] = location
+                        break
+
+        lines_to_auto_wave = self.filtered(lambda l: l.picking_type_id.wave_grouping and not l.batch_id.is_wave and
+                                        (l.picking_type_id.should_group_waves == 'partially_available' or l.state == 'assigned'))
+        lines_to_auto_wave = lines_to_auto_wave.filtered(lambda l: not l.picking_type_id.wave_group_by_category or l.product_id.categ_id in l.picking_type_id.wave_category_ids)
+        lines_to_auto_wave = lines_to_auto_wave.filtered(lambda l: not l.picking_type_id.wave_group_by_location or lines_nearest_parent_locations[l])
+
+        possible_waves_by_line = lines_to_auto_wave._get_possible_waves(nearest_parent_locations=lines_nearest_parent_locations)
+        lines_by_wave = defaultdict(lambda: self.env['stock.move.line'])
+        for line, possible_waves in possible_waves_by_line.items():
```

Comment:
Nice effort!

Add new method to stock_move_line:
```
def _is_auto_wave_batchable(self):
    self.ensure_one()
    if not self.picking_id \
       or self.state not in ['draft', 'done', 'cancel'] \
       or not self.quantity \
       or self.batch_id.is_wave \
       or not self.picking_type_id.is_auto_wave_grouped() \
       or (self.picking_type_id.wave_group_by_category and self.product_id.categ_id not in self.picking_type_id.wave_category_ids)
       :
        return False
    return True
```
Note that I dropped the `should_group_waves == 'partially_available'` on purpose because I think this check/value doesn't make sense. The wave grouping occurs at the move line (ml) level, but the `ml.state = ml.move_id.state`, which means that a ml with `state = partially_available` can have nothing reserved or done, so it shouldn't be grouped into a wave. I will discuss this with Thomas in person.

Also note that I did the painful multiple `or` lines because once the first "True" is hit, the rest of the checks are skipped

```suggestion
        lines_nearest_parent_locations = defaultdict(lambda: self.env['stock.location'])
        batchable_line_ids = OrderedSet()
        for line in self:
            if  not line._is_auto_wave_batchable()
                continue
            if not line.picking_type_id.wave_group_by_location:
                batchable_line_ids.add(line.id)
                continue
            if 
                for location in line.picking_type_id.wave_location_ids[::-1]:
                    if line.location_id._child_of(location):
                        lines_nearest_parent_locations[line] = location
                        batchable_line_ids.add(line.id)
                        break
       batchable_lines = self.env['stock.move.line'].browse(batchable_line_ids)
        possible_waves_by_line = batchable_lines._get_possible_waves(nearest_parent_locations=lines_nearest_parent_locations)
        lines_by_wave = defaultdict(lambda: self.env['stock.move.line'])
        for line, possible_waves in possible_waves_by_line.items():
```

note that the location_ids already have `_order = 'complete_name, id'` set on stock.location (i.e. it's already ordered by complete_name) and so we can loop through it in reverse thanks to python magic.

Also note that my brain is now dead. I will continue reviewing the rest of the auto-wave batching logic tomorrow :skull: